### PR TITLE
keyprovider: Added support for ttrpc protocol

### PR DESCRIPTION
**Pull request by pravinrajr9**
ttrpc protocol implementation for keyprovider

### CODE REVIEW

**Comment by jiangliu** [comment]
The grpc stack is too heavy, and it's really great to switch to ttrpc:)

**Comment by sameo** [comment]
@pravinrajr9 Would you mind rebasing that PR, please?

**Comment by pravinrajr9** [comment]
> @pravinrajr9 Would you mind rebasing that PR, please?

sure

**Comment by pravinrajr9** [comment]
@sameo @arronwy please review

File: build.rs [hunk 1]
```diff
@@ -1,11 +1,57 @@
...
+    let protos = vec!["src/utils/proto/keyprovider.proto"];
     #[cfg(feature = "keywrap-keyprovider")]
     tonic_build::configure()
         .build_server(true)
-        .out_dir("src/utils/")
-        .compile(&["src/utils/proto/keyprovider.proto"], &["src/utils"])?;
+        .out_dir("src/utils/grpc")
+        .compile(&protos, &["src/utils"])?;
+
+    ttrpc_codegen::Codegen::new()
```
Comment (arronwy) [inline build.rs:16]: Can we add a new feature for ttrpc support? As a keyprovider client, we may need support only one type with build config.

File: build.rs [hunk 2]
```diff
@@ -1,11 +1,57 @@
...
+        .include("src/utils")
+        .rust_protobuf()
+        .customize(Customize {
+            async_all: true,
+            ..Default::default()
+        })
+        .run()
+        .expect("Gen code failed.");
+
+    // To fix clippy warnings of code generated from ttrpc_codegen
```
Comment (arronwy) [inline build.rs:28]: Can we use `#[rustfmt::skip]` to skip the fmt check for generated `keyprovider_ttrpc` module?

Comment (pravinrajr9) [reply build.rs:28]: No these are not rustfmt errors, these are something else, you can check here https://github.com/containerd/ttrpc-rust/blob/master/example/build.rs

File: src/keywrap/keyprovider.rs [hunk 3]
```diff
@@ -135,6 +142,44 @@ impl KeyProviderKeyWrapProtocolOutput {
...
+                .await
+                .map_err(|_| {
+                    anyhow!(
+                        "keyprovider: Error from ttrpc server for {:?} operation",
+                        OpKey::Wrap.to_string()
+                    )
+                })?,
+
+            OpKey::Unwrap => kc1
+                .un_wrap_key(context::with_timeout(20 * 1000 * 1000), &req)
```
Comment (jodh-intel) [inline src/keywrap/keyprovider.rs:166]: This timeout is repeated above. Please create a global `const` for it.

File: src/keywrap/keyprovider.rs [hunk 4]
```diff
@@ -135,6 +142,44 @@ impl KeyProviderKeyWrapProtocolOutput {
...
+        let c = Client::connect(conn).unwrap();
+
+        let kc = keyprovider_ttrpc::KeyProviderServiceClient::new(c);
+        let kc1 = kc.clone();
+        let mut req = keyprovider_ttrpc_structs::keyProviderKeyWrapProtocolInput::new();
+        req.set_KeyProviderKeyWrapProtocolInput(input);
+
+        let ttrpc_output = match operation {
+            OpKey::Wrap => kc1
+                .wrap_key(context::with_timeout(20 * 1000 * 1000), &req)
```
Comment (jodh-intel) [inline src/keywrap/keyprovider.rs:156]: Could you define a global `const` for this duplicated timeout value (with a name that shows the units for this value).

File: src/keywrap/keyprovider.rs [hunk 5]
```diff
@@ -135,6 +142,44 @@ impl KeyProviderKeyWrapProtocolOutput {
         serde_json::from_slice(&resp_bytes)
             .map_err(|_| anyhow!("keyprovider: failed to deserialize message from binary executor"))
     }
+
+    async fn from_ttrpc(input: Vec<u8>, conn: &str, operation: OpKey) -> Result<Self> {
+        let c = Client::connect(conn).unwrap();
```
Comment (jodh-intel) [inline src/keywrap/keyprovider.rs:147]: Please use `?` rather than `unwrap()` as this function returns a `Result`.

**Review by jodh-intel — COMMENTED** [review]
Thanks @pravinrajr9 - a few comments...

**Comment by jiangliu** [comment]
Sorry for late reply, this is what I really appreciate:)
We should use ttrpc instead of grpc to reduce the dependency chain.
https://github.com/confidential-containers/guest-components/issues/257

**Comment by arronwy** [comment]
Closed with rebased version: https://github.com/confidential-containers/ocicrypt-rs/pull/45

**Pull request closed without merging**